packages/react/src/streams.ts

```diff
@@ -0,0 +1,47 @@
+import { useSyncExternalStore } from 'react'
+
+export type Stream<V> = {
+  get: () => V
+  subscribe: (handler: (value: V) => void) => () => void
+}
+
+export type DataSourceStateStream = Stream<{
+  status: 'connecting' | 'connected' | 'closed' | 'closing' | 'error' | 'undetermined'
```

Comment:
These are fine for now but not great. What if you are `closed` but you also have an error? Also, what is the purpose of `undetermined`? Does that basically mean "closed"? Eventually maybe error should be in a separate field

Also the opposite of "connected" is "disconnected", not "closed"

Also I'm not sure "closing" will ever be used. You should be able to instantly close the connection from the client

Reply:
Yeah,

"undetermined" is carry over from previous work. I would remove it, as it can't happen in the WebSocket connection.

All other statuses are directly taken from WebSocket. However, I think to further simplify this, we should just make it "connected" and "disconnected" (or "closed", to unify with WebSocket spec), because these are the only statuses that we can listen to (via "connected" and "closed" events). All other statuses, while being part of the spec, are internal and we can't really get them.

"error" should also be removed. You can have an error and still be connected, but in case it throws, it will just change status to disconnected.

WDYT?

Reply:
Yep! Error could be a different stream

Reply:
Simplified to just "connected" and "disconnected". It can also be undefined, which means not connected and not disconnected (default state).